### DISCUSSION TITLE
Multithreading

### DISCUSSION THREAD

**Discussion by sa-dd** [Rapptz/discord.py#9749] [Q&A]
So I've been trying to make this discord bot that uses yolov8 object detection model for inference. I'm running the inference function inside the `on_message` event, but due to it's blocking nature, I can't make more than one request at a time. I was able to make multiple threads for the model instance, and was able to deal with multiple requests in parallel. Now the issue I'm facing is with `message.channel.send` as it is a coroutine, but threads are only able to run synchronous functions. I was able to solve the messaging as well by using `asyncio.run_coroutine_threadsafe()` but only when I exclude `discord.ui.View()`. 

code: 
```
asyncio.run_coroutine_threadsafe(message.channel.send(f"{message.author.mention} Here are your results: ",file= discord.File(fp=image_bytesio, filename="image.jpeg"), embeds=sendEmbed, view = discord.ui.View().add_item(discord.ui.Button(label="Join My Discord Server", emoji=emoji,style=discord.ButtonStyle.gray, url="https://discord.gg/server"))), client.loop)
```
error:
```
Exception in thread Thread-92 (detect):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "<ipython-input-36-5d42097b2506>", line 245, in detect
  File "/usr/local/lib/python3.10/dist-packages/discord/ui/view.py", line 192, in __init__
    self.__stopped: asyncio.Future[bool] = asyncio.get_running_loop().create_future()
RuntimeError: no running event loop
```
It works fine, when I don't add it. I checked the source code, and I think it's not being able to detect the loop. Is there a way, I can pass the running loop to it?

**Comment by ika2kki** [comment]
hi sorry for the late response~
when you construct a `ui.View`, it expects an asyncio event loop running in the current thread (which should only be the main one) for internal behaviour. so, you need to make one within the main thread.
what you can do is layer this code under another coroutine so it all happens under the event loop, eg.
```py
async def send_results():
    msg = await message.channel.send(f"{message.author.mention} Here are your results: ",file= discord.File(fp=image_bytesio, filename="image.jpeg"), embeds=sendEmbed, view = discord.ui.View().add_item(discord.ui.Button(label="Join My Discord Server", emoji=emoji,style=discord.ButtonStyle.gray, url="https://discord.gg/server")))
    return msg

asyncio.run_coroutine_threadsafe(send_results(), client.loop)
```

**Reply by sa-dd** [reply]
Hey, thanks for the reply, but I already solved the issue. Someone advised me to use `asyncio.run_in_executor` from my async functions, and not try to start async functions from threads, that are not asynchronous (which is what call_soon_threadsafe attempts). So I just make new threads from inside the `on_message`, and return all the data, while attaching the ui.View object from inside the `on_message`, before sending the message.

**Reply by ika2kki** [reply]
ya thats an ideal scenario. gj!
